File: udp/UDPClient.java
package udp;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.net.DatagramPacket;
import java.net.DatagramSocket;
import java.net.InetAddress;
import java.net.SocketAddress;
import java.nio.charset.StandardCharsets;

public class UDPClient {

  private static final UDPClient CLIENT = new UDPClient();


  private UDPClient() {
  }

  public static void main(String[] args) {
    CLIENT.receive();
    CLIENT.send();
  }

  private void send() {
    Thread thread = new Thread(() -> {
      BufferedReader bis = new BufferedReader(new InputStreamReader(System.in));
      try {
        DatagramSocket socket = new DatagramSocket();
        InetAddress inetAddress = InetAddress.getByName("192.168.42.255");
        String sendMessage;
        while (!(sendMessage = bis.readLine()).equals("exit")) {
          DatagramPacket send = new DatagramPacket(sendMessage.getBytes(StandardCharsets.UTF_8),
              sendMessage.length(), inetAddress, 8888);
          socket.send(send);
        }
        socket.close();
      } catch (IOException e) {
        e.printStackTrace();
      }
    });
    thread.start();
  }

  private void receive() {

    Thread thread = new Thread(() -> {
      try {
        DatagramSocket socket = new DatagramSocket(8888);
        byte[] buffer;
        while (true) {
          buffer = new byte[1024];
          DatagramPacket receive = new DatagramPacket(buffer, buffer.length);
          socket.receive(receive);
          byte[] get = receive.getData();
          String getString = new String(get, 0, get.length).trim();
          SocketAddress client = receive.getSocketAddress();
          System.out.println(client + "--------------" + getString);
        }
      } catch (IOException e) {
        e.printStackTrace();
      }
    });
    thread.setDaemon(true);
    thread.start();
  }
}

File: udp/UDPClient.kt
package udp

import java.io.BufferedReader
import java.io.IOException
import java.io.InputStreamReader
import java.net.DatagramPacket
import java.net.DatagramSocket
import java.net.InetAddress
import java.nio.charset.StandardCharsets

class UDPClient private constructor() {

    private fun send() {
        val thread = Thread {
            val bis = BufferedReader(InputStreamReader(System.`in`))
            try {
                val socket = DatagramSocket()
                val inetAddress = InetAddress.getByName("192.168.43.255")
                var sendMessage: String
                while ("exit" != bis.readLine().also { sendMessage = it }) {
                    val send = DatagramPacket(sendMessage.toByteArray(StandardCharsets.UTF_8),
                            sendMessage.length, inetAddress, 8888)
                    socket.send(send)
                }
                socket.close()
            } catch (e: IOException) {
                e.printStackTrace()
            }
        }
        thread.start()
    }

    private fun receive() {

        val thread = Thread {
            try {
                val socket = DatagramSocket(8888)
                var buffer: ByteArray
                while (true) {
                    buffer = ByteArray(1024)
                    val receive = DatagramPacket(buffer, buffer.size)
                    socket.receive(receive)
                    val get = receive.data
                    val getString = String(get, 0, get.size).trim { it <= ' ' }
                    val client = receive.socketAddress
                    println("$client--------------$getString")
                }
            } catch (e: IOException) {
                e.printStackTrace()
            }
        }
        thread.isDaemon = true
        thread.start()
    }

    companion object {

        private val CLIENT = UDPClient()

        @JvmStatic
        fun main(args: Array<String>) {
            CLIENT.receive()
            CLIENT.send()
        }
    }
}

